### PR TITLE
Workdamage

### LINKED ISSUE #72: Integrate work-based damage model
I want to have a generic work-based damage model for creep-fatigue in NEML.

### PR DESCRIPTION
**Pull request by reverendbedford**
Integrated the work-based damage model into NEML.

Closes #72.

### CODE REVIEW

File: src/interpolate.cxx [hunk 1]
```diff
@@ -387,6 +477,45 @@ double ExpInterpolate::derivative(double x) const
...
+{
+  return neml::make_unique<PowerLawInterpolate>(
+      params.get_parameter<double>("A"),
+      params.get_parameter<double>("B")
+      ); 
+}
+
+double PowerLawInterpolate::value(double x) const
+{
+  return A_ * pow(x, B_);
```
Comment (arovinelli) [inline src/interpolate.cxx:511]: `pow `->`std:: pow `

File: src/interpolate.cxx [hunk 2]
```diff
@@ -387,6 +477,45 @@ double ExpInterpolate::derivative(double x) const
...
+}
+
+double PowerLawInterpolate::derivative(double x) const
+{
+  return A_ * B_ * pow(x, B_-1.0);
```
Comment (arovinelli) [inline src/interpolate.cxx:516]: `pow` ->`std:: pow`

File: src/interpolate.cxx [hunk 3]
```diff
@@ -311,6 +311,96 @@ double PiecewiseLogLinearInterpolate::derivative(double x) const
   }
 }
 
+PiecewiseSemiLogLinearInterpolate::PiecewiseSemiLogLinearInterpolate(
```
Comment (arovinelli) [inline src/interpolate.cxx:314]: I would change the name to `PiecewiseSemiLogXLinearInterpolate` as `y` is assumed to be in the right dimensions

File: src/interpolate.h [hunk 4]
```diff
@@ -124,6 +126,31 @@ class NEML_EXPORT PiecewiseLogLinearInterpolate: public Interpolate {
 
 static Register<PiecewiseLogLinearInterpolate> regPiecewiseLogLinearInterpolate;
 
+/// Piecewise semiloglinear interpolation
+class NEML_EXPORT PiecewiseSemiLogLinearInterpolate: public Interpolate {
```
Comment (arovinelli) [inline src/interpolate.h:130]: I would change the name to `PiecewiseSemiLogXLinearInterpolate`

File: test/test_interpolate.py [hunk 5]
```diff
@@ -110,8 +110,26 @@ def test_interpolate(self):
     yp[xs < self.validx[0]] = np.log(self.points[0])
     yp[xs > self.validx[-1]] = np.log(self.points[-1])
     ys2 = np.exp(yp)
-    print(ys1)
-    print(ys2)
+    self.assertTrue(np.allclose(ys1, ys2))
+
+class TestPiecewiseSemiLogLinearInterpolate(unittest.TestCase, BaseInterpolate):
```
Comment (arovinelli) [inline test/test_interpolate.py:115]: some thing with the name I here`PiecewiseSemiLogXLinearInterpolate`

File: doc/sphinx/damage/scalar.rst [hunk 6]
```diff
@@ -14,7 +14,7 @@ It implements the stress update function
       \bm{\varepsilon}_{n+1}, \bm{\varepsilon}_{n},
       T_{n+1}, T_{n},
       t_{n+1}, t_{n},
-      \bm{\sigma}_{n} / (1 - \omega_{n+1},
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply doc/sphinx/damage/scalar.rst:17]: Yes, this was a typo I guess I randomly fixed back in April.  Revision also fixes the alpha_n issue right next to it.

File: doc/sphinx/damage/scalar.rst [hunk 7]
```diff
@@ -14,7 +14,7 @@ It implements the stress update function
...
+      \bm{\sigma}_{n} / (1 - \omega_{n}),
       \bm{\alpha}_{n},
       \bm{\alpha}_{n},
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply doc/sphinx/damage/scalar.rst:19]: Yep, see above.  One of these should be \alpha_{n+1}

File: neml/drivers.py [hunk 8]
```diff
@@ -1330,7 +1331,8 @@ def creep(model, smax, srate, hold, T = 300.0, nsteps = 250,
   return {'time': np.copy(time), 'strain': np.copy(strain),
       'stress': np.copy(stress), 'rtime': np.copy(rtime[:-1]),
       'rrate': np.copy(rrate), 'rstrain': np.copy(rstrain[:-1]),
-      'tstrain': np.copy(strain[ri:-1]), 'failed': failed}
+      'tstrain': np.copy(strain[ri:-1]), 'failed': failed,
+      'history': np.array(driver.stored_int)}
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply neml/drivers.py:1335]: Sure.

File: src/damage.cxx [hunk 9]
```diff
@@ -121,7 +121,12 @@ int NEMLScalarDamagedModel_sd::init_x(double * const x, TrialState * ts)
 {
   SDTrialState * tss = static_cast<SDTrialState *>(ts);
   std::copy(tss->s_n, tss->s_n+6, x);
-  x[6] = tss->w_n;
+  if (tss->w_n == 0.0) {
+    x[6] = d_guess();
+  }
+  else {
+    x[6] = tss->w_n;
+  }
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply src/damage.cxx:129]: Yes, it's a way to allow the models to provide a guess for t = 0.  I added a comment.

File: src/damage.cxx [hunk 10]
```diff
@@ -1416,6 +1421,211 @@ double NEMLStandardScalarDamagedModel_sd::dep(
...
+    std::shared_ptr<NEMLModel_sd> base,
+    std::shared_ptr<Interpolate> alpha,
+    double tol, int miter,
+    bool verbose, bool truesdell,
+    double eps) :
+      NEMLScalarDamagedModel_sd(elastic, base, alpha, tol, miter, 
+                                        verbose, truesdell, false, 0, 1), 
+      Wcrit_(Wcrit), n_(n), eps_(eps)
+{
+
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply src/damage.cxx:1439]: k

File: src/damage.cxx [hunk 11]
```diff
@@ -1416,6 +1421,211 @@ double NEMLStandardScalarDamagedModel_sd::dep(
...
+                   const double * const e_np1, const double * const e_n,
+                   const double * const s_np1, const double * const s_n,
+                   double T_np1, double T_n,
+                   double t_np1, double t_n,
+                   double * const dd) const
+{
+  if (d_np1 <= 0.0) {
+    std::fill(dd, dd+6, 0.0);
+    return 0;
+  }
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply src/damage.cxx:1540]: No, just for numerics.  The damage variable is part of the nonlinear system so Newton's method can provide junk (like d < 0) for uncoverged iterations.  I added a comment.

File: src/damage.cxx [hunk 12]
```diff
@@ -1416,6 +1421,211 @@ double NEMLStandardScalarDamagedModel_sd::dep(
...
+double NEMLWorkDamagedModel_sd::workrate(
+    const double * const strain_np1, const double * const strain_n,
+    const double * const stress_np1, const double * const stress_n,
+    double T_np1, double T_n, double t_np1, double t_n,
+    double d_np1, double d_n) const
+{
+  double dt = t_np1 - t_n;
+  if (dt <= 0.0) {
+    return 0.0;
+  }
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply src/damage.cxx:1606]: It's really for dt == 0 :)  But == with floating points is bad.

File: src/interpolate.cxx [hunk 13]
```diff
@@ -311,6 +311,96 @@ double PiecewiseLogLinearInterpolate::derivative(double x) const
...
+{
+  if (x <= points_.front()) {
+    return 0.0;
+  }
+  else if (x >= points_.back()) {
+    return 0.0;
+  }
+  else {
+    auto it = points_.begin();
+    for (; it != points_.end(); ++it) {
```
[the remark this answers is not in the document]
Comment (reverendbedford) [reply src/interpolate.cxx:391]: Nope, can't.  Need access to "it" after the end of the loop.

**Pull request merged**